### PR TITLE
Use AllowanceManager in ZeroExInteraction

### PR DESCRIPTION
**Pull request by MartinquaXD**
There were quite a few [simulation failures](https://cowservices.slack.com/archives/C0361CDD1FZ/p1661543463263019) for settlements including `ZeroExInteraction`s. This was caused by the `ZeroExInteraction` not automatically appending the appropriate `Approval` interaction when the `ZeroExInteraction` was used and encoded in a settlement.

To fix that I followed how the balancer liquidity and interaction handles this and implemented it accordingly for ZeroEx.

### Test Plan
added unit test verifying the encoded `Approval` and `ZeroExInteraction`

### CODE REVIEW

**Review by bh2smith — COMMENTED** [review]
Great stuff here! My only question is whether or not this approved every time or only when necessary.

**Comment by MartinquaXD** [comment]
> Great stuff here! My only question is whether or not this approved every time or only when necessary.

When the allowance is big enough for the interaction the `AllowanceManager` issues an `Approval::AllowanceSufficient` interaction which encodes itself into an empty vector (as the call_data) making it a no-op.

File: crates/solver/src/liquidity/zeroex.rs
Comment (josojo) [inline crates/solver/src/liquidity/zeroex.rs:186]: What would happen if we approved already zeroEx.address() with 5 tokens. And we have 2 different orders each selling 4 tokens? Then I assume we would get always Allowance::Sufficient and not set an approval, but the tx would fail as the approval is not sufficient.

I assume that it's not really an issue, as we assume that we have an infinitive approval set, right?

**Review by josojo — APPROVED** [review]
Like it! Thanks for the super quick fix

File: crates/solver/src/liquidity/zeroex.rs
Comment (MartinquaXD) [reply crates/solver/src/liquidity/zeroex.rs:186]: AFAICS you are correct. I think all the liquidity implementations currently have this problem but it can be fixed for all by a patch in the `Allowances` implementation. Every time we get the approval for a token we could immediately deduct the approved amount and if a later interaction would put the allowance over the edge an approval interaction would be issued.

File: crates/driver/src/main.rs
Comment (nlordell) [inline crates/driver/src/main.rs:387]: Should we pass around a `Arc<dyn AllowanceManaging>` instead of creating a new one in each sub-component?

**Review by nlordell — APPROVED** [review]

File: crates/driver/src/main.rs
Comment (MartinquaXD) [reply crates/driver/src/main.rs:387]: Good point! 👍

Comment (MartinquaXD) [reply crates/driver/src/main.rs:387]: I will do that in a follow up PR.

**Pull request merged**
